### PR TITLE
Activate Complexity and Tractability chapter

### LINKED ISSUE #381: Update time calculator interactive in tractibility chapter
http://www.csfieldguide.org.nz/releases/1.9.9/_static/widgets/tract-scaling-v2.html

This interactive needs to be updated to include missing complexities for searching algorithms (n and log(n)), particularly as it is currently linked to from the searching algorithms ncea assessment guide.

Possibly related to recreating big number calculator, #23

### LINKED ISSUE #722: Add missing complexity and tractability sections
Currently several sections are not loaded in the configuration file.

### PR DESCRIPTION
**Pull request by courtneycb**
This PR activates the Complexity and Tractability chapter and includes:

- 2 new interactives, City Trip and Algorithm Timer.
- Grammar fixes

Please checkout the branch and play around with the interactives. In particular City Trip.

Resolves #381 by adding in the logarithmic and linear complexities for the Algorithm Timer interactive and resolves #722 as the sections have been added back in. 

There seems to be an issue with the interactive thumbnail for City Trip as it can't be found. I have commented out the tests for interactive thumbnails (advised by Jack) and created an issue for these to be added back in. See #919

### CODE REVIEW

File: csfieldguide/chapters/content/en/complexity-and-tractability/sections/tractability.md [hunk 1]
```diff
@@ -16,7 +16,7 @@ If you are interested, you can [download the spreadsheet here](files/tractable-s
 There's a very simple rule that computer scientists use to decide if an algorithm is tractable or not, based on the complexity (estimated number of steps) of the algorithm.
 Essentially, if the algorithm takes an exponential amount of time or worse for an input of size *n*, it is labelled as intractable.
 This simple rule is a bit crude, but it's widely used and provides useful guidance.
-(Note that a factorial amount of time, *n!*, is intractable because it's bigger than an exponential function.)
+(Note that a factorial amount of time, *n!*, is intractable because it's bigger than an exponential function).
 
 To see what this means, let's consider how long various algorithms might take to run.
 The following interactive will do the calculations for you to estimate how long an algorithm might take to run.
```
Comment (eAlasdair) [inline csfieldguide/chapters/content/en/complexity-and-tractability/sections/tractability.md]: This, and sentences elsewhere including the above teacher note, appear to refer to the interactive that was deleted further down. If the interactive is no longer there the text should be updated to reflect that

File: csfieldguide/templates/interactives/city-trip.html [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    <div class="row justify-content-center">
+      <div class="col-sm-12 col-lg-8">
+        <h3 class="text-center my-3">{% trans "City Trip" %}</h3>
+      </div>
+    </div>
+
+    <div class="row mb-2 justify-content-left">
+      <div class="col-sm-12 col-lg-2 mt-1">
+        <label>
+          {% trans "Number of cities: <span id='num-cities'>3</span>" %}
```
[the remark this answers is not in the document]
Comment (eAlasdair) [reply csfieldguide/templates/interactives/city-trip.html]: i.e. go `{% trans "Number of cities:" %} <span id='num-cities'>3</span>`
-----------------------------------------------------^ this is the space that's the 'last character'

[the remark this answers is not in the document]
Comment (courtneycb) [reply csfieldguide/templates/interactives/city-trip.html]: Sweet. I'm going to leave a trailing space in some areas where it messes with the code style. (e.g line 85).

File: csfieldguide/chapters/content/en/complexity-and-tractability/sections/the-travelling-salesman-problem.md [hunk 3]
```diff
@@ -100,15 +100,15 @@ Just like in that activity you listed all the possible ordering for the values i
 How many possible routes are there for the larger example you have generated?
 How is this related to permutation sort, and factorials?
 How long would it take to calculate the shortest route in your map, assuming the computer can check 1 billion (1,000,000,000) possible routes per second?
-(i.e. it can check one route per nanosecond) What can you conclude about the cost of this algorithm?
+(i.e. it can check one route per nanosecond). What can you conclude about the cost of this algorithm?
```
[the remark this answers is not in the document]
Comment (courtneycb) [reply csfieldguide/chapters/content/en/complexity-and-tractability/sections/the-travelling-salesman-problem.md]: I removed '(i.e. it can check one route per nanosecond)'. I thought it was unnecessary as we already stated it can check 1 billion routes per second. Thoughts?

[the remark this answers is not in the document]
Comment (courtneycb) [reply csfieldguide/chapters/content/en/complexity-and-tractability/sections/the-travelling-salesman-problem.md]: I also kept 1,000,000,000 as I feel it's good to have as a copy-paste for the interactive

File: csfieldguide/chapters/content/en/complexity-and-tractability/sections/the-travelling-salesman-problem.md [hunk 4]
```diff
@@ -100,15 +100,15 @@ Just like in that activity you listed all the possible ordering for the values i
...
 Would this be a good way for the cray fisher to decide which path to take?
```
[the remark this answers is not in the document]
Comment (courtneycb) [reply csfieldguide/chapters/content/en/complexity-and-tractability/sections/the-travelling-salesman-problem.md]: Do you have a suggestion for a replacement? I considered fisherman but not sure what our stance is on using words with '-man' in them. I googled it and apparently people tend to use 'fisher' now but it sounds strange to me.

File: csfieldguide/static/interactives/algorithm-timer/js/algorithm-timer.js [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+const COMPLEXITY_TEXT = {
+  'squared': 'n&sup2;',
+  'cubed': 'n&sup3;',
+  'fourth-power': 'n<sup>4</sup>',
+  'sixth-power': 'n<sup>6</sup>',
+  'exponential': '2<sup>n</sup>',
+  'factorial': 'n!',
+};
+
+const TIME_SCALERS = {
```
[the remark this answers is not in the document]
Comment (courtneycb) [reply csfieldguide/static/interactives/algorithm-timer/js/algorithm-timer.js]: I've made them relative to each other, it's just in seconds not milliseconds.

File: csfieldguide/chapters/content/en/complexity-and-tractability/sections/the-travelling-salesman-problem.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (eAlasdair) [reply csfieldguide/chapters/content/en/complexity-and-tractability/sections/the-travelling-salesman-problem.md]: That also works fine, I approve

File: csfieldguide/chapters/content/en/complexity-and-tractability/sections/the-travelling-salesman-problem.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (eAlasdair) [reply csfieldguide/chapters/content/en/complexity-and-tractability/sections/the-travelling-salesman-problem.md]: Our stance, according to the docs, is to use gender neutral terms when referring to a hypothetical person, but that is very ambiguous. Since: people who catch crayfish don't seem to have their own term; 'man' itself (even before you add other characters) is historically gender neutral; and noone's going to make too much of a fuss over a basic example, I think it's okay to say fisherman

File: csfieldguide/templates/interactives/city-trip.html (hunk 2 above)
[the remark this answers is not in the document]
Comment (eAlasdair) [reply csfieldguide/templates/interactives/city-trip.html]: That's fine

File: csfieldguide/chapters/content/en/complexity-and-tractability/sections/the-travelling-salesman-problem.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (courtneycb) [reply csfieldguide/chapters/content/en/complexity-and-tractability/sections/the-travelling-salesman-problem.md]: We had an office discussion and a quick Google - apparently crayfisher is the right term so will go with that.

[the remark this answers is not in the document]
Comment (eAlasdair) [reply csfieldguide/chapters/content/en/complexity-and-tractability/sections/the-travelling-salesman-problem.md]: Fine with me

**Comment by eAlasdair** [comment]
Since the Licence style from #863 has been approved by Jack, we should start using it in all current PRs

File: LICENCE-THIRD-PARTY [hunk 6]
```diff
@@ -32,6 +59,15 @@ licensed under MIT licence.
 third-party-licences/Detectorjs.txt
 ==============================================================================
 
+==============================================================================
+Mathjs
+------------------------------------------------------------------------------
+https://github.com/josdejong/mathjs
+Copyright 2013-2019 Jos de Jong 
+Licensed under the Apache License, Version 2.0
```
Comment (eAlasdair) [inline LICENCE-THIRD-PARTY]: 'licenced' is lowercase everywhere else - I can't remember why I made it that way, did CSU do that?

File: csfieldguide/chapters/content/en/complexity-and-tractability/sections/the-travelling-salesman-problem.md [hunk 7]
```diff
@@ -2,25 +2,25 @@
...
 
 Current algorithms for finding the optimal TSP solution aren't a lot better than simply trying out all possible paths through the map (as in the interactive at the start of this chapter).
 The number of possible paths gets out of hand; it's an intractable approach.
 In the project below you'll be estimating how long it would take.
 
 While TSP was originally identified as being the problem that sales people face when driving to several different locations and wanting to visit them in the order that leads to the shortest route (less petrol usage), the same problem applies to many other situations as well.
-Courier and delivery companies have variants of this problem — often with extra constraints such as limits on how long a driver can work for, or allowing for left hand turns being faster than right-hand ones (in NZ at least!)
+Courier and delivery companies have variants of this problem &ndash; often with extra constraints such as limits on how long a driver can work for, or allowing for left hand turns being faster than right-hand ones (in NZ at least!)
```
Comment (eAlasdair) [inline csfieldguide/chapters/content/en/complexity-and-tractability/sections/the-travelling-salesman-problem.md]: Need a full stop at the end of the sentence.
Some might not recognise 'NZ', maybe we could generalise it to be about left-side traffic instead - or change the example to be understood regardless of traffic direction

File: csfieldguide/chapters/content/en/complexity-and-tractability/sections/the-travelling-salesman-problem.md [hunk 8]
```diff
@@ -140,27 +140,27 @@ Show a map where they would be different (you can choose where to place the cray
 
 For your larger map, show why you don’t have an optimal solution.
 The best way of doing this is to show a route that is similar to, but shorter than the approximate solution.
-The shorter solution you find doesn’t have to be the optimal solution, it just has to be shorter than the one identified by the approximate algorithm (Talk to your teacher if you can’t find a shorter route and they will advise on whether or not you should generate a new map).
+The shorter solution you find doesn’t have to be the optimal solution, it just has to be shorter than the one identified by the approximate algorithm. Talk to your teacher if you can’t find a shorter route and they will advise on whether or not you should generate a new map.
```
Comment (eAlasdair) [inline csfieldguide/chapters/content/en/complexity-and-tractability/sections/the-travelling-salesman-problem.md]: New sentence on a new line

File: csfieldguide/chapters/content/en/complexity-and-tractability/sections/tractability.md [hunk 9]
```diff
@@ -52,7 +52,7 @@ It's still completely impractical!
 
 Another way to solve problems faster is to have multiple processors work on different solutions at the same time.
 If you were to buy 1,000 processors (e.g. 1,000 computers, or 250 4-core computers) and have each one test out different routes, then the solution could be found 1,000 times faster.
-Try changing the number of processors to 1,000, and see how long that would take (you may need to change the units back — is it seconds? hours? days?)
+Try changing the number of processors to 1,000 and see how long that would take (you may need to change the units back). Is it seconds? Hours? Days?
```
Comment (eAlasdair) [inline csfieldguide/chapters/content/en/complexity-and-tractability/sections/tractability.md]: `Is it seconds? Hours? Days?` can stay on the same line, but it should be on a new one

File: csfieldguide/static/interactives/city-trip/README.md [hunk 10]
```diff
@@ -0,0 +1,5 @@
+# City Trip Interactive
+
+**Author:** Courtney Bracefield
+
+This interactive is created to demonstrate the travelling salesman problem.
```
Comment (eAlasdair) [inline csfieldguide/static/interactives/city-trip/README.md]: I think you could flesh out both of these a bit - How does it demonstrate TSP? What can the user do? What third-party libraries were used here and how can I find their licences?

File: LICENCE-THIRD-PARTY (hunk 6 above)
Comment (eAlasdair) [reply LICENCE-THIRD-PARTY]: Sorry, 'licensed' - as its a verb

File: csfieldguide/static/interactives/city-trip/README.md (hunk 10 above)
Comment (eAlasdair) [reply csfieldguide/static/interactives/city-trip/README.md]: Also this interactive existed before you recreated it, so mention the original author somewhere

File: csfieldguide/templates/interactives/city-trip.html [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+          <div class="x-tick zero"><span>0</span></div>
+          <div class="x-tick"><span>1</span></div>
+          <div class="x-tick"><span>2</span></div>
+          <div class="x-tick"><span>3</span></div>
+          <div class="x-tick"><span>4</span></div>
+          <div class="x-tick"><span>5</span></div>
+          <div class="axis-untis"><span>km</span></div>
+        </div>
+        <div class="text-center mt-5">
+          <p class="mb-1">{% trans "Trial route: " %}<span id='trial-route'>0</span></p>
```
Comment (eAlasdair) [inline csfieldguide/templates/interactives/city-trip.html]: You could move the space as I mentioned earlier here:
`<p class="mb-1">{% trans "Trial route:" %} <span id='trial-route'>0</span></p>`

File: csfieldguide/static/interactives/city-trip/js/city-trip.js [hunk 12]
```diff
@@ -0,0 +1,642 @@
...
+
+/** Calculates how many years,months,days.. etc there are in runningTimeLeft.
+ *  runningTimeLeft is given in seconds. */
+function formatTime(runningTimeLeft) {
+  var years = Math.floor(runningTimeLeft / 31536000);
+  var months = Math.floor((runningTimeLeft % 31536000) / 2628000);
+  var days = Math.floor((runningTimeLeft % 2628000) / 86400);
+  var hours = Math.floor((runningTimeLeft % 86400) / 3600);
+  var minutes = Math.floor((runningTimeLeft % 3600) / 60);
+  var seconds = (runningTimeLeft % 60).toFixed(1);
```
Comment (eAlasdair) [inline csfieldguide/static/interactives/city-trip/js/city-trip.js]: Lots of magical mystical fairy numbers here, It's simple enough to tell it's set in seconds, but perhaps explain the first modulo - 'we only want to base it from the remainder' or something

File: csfieldguide/static/interactives/city-trip/js/city-trip.js [hunk 13]
```diff
@@ -0,0 +1,642 @@
...
+cytoscape.use(automove);
+var stopPathFinding = false;
+var stoppedMidExecution = false; // true if the stop button been clicked mid path finding.
+
+const STATUS_COMPLETE = gettext("complete!");
+const STATUS_READY = gettext("ready to go!");
+const STATUS_STOPPED = gettext("stopped");
+const STATUS_RUNNING = gettext("running");
+
+$(document).ready(function() {
```
Comment (eAlasdair) [inline csfieldguide/static/interactives/city-trip/js/city-trip.js]: I think we can do better for phone screens

image

File: csfieldguide/static/interactives/city-trip/js/city-trip.js [hunk 14]
```diff
@@ -0,0 +1,642 @@
...
+   *  Non-recursive version used so we can draw the new path as soon as it is generated.
+   */
+async function permutations(cy, cy2, intermediateCities, bestRouteDistance, startingCity, k) {
+  // c is an encoding of he stack state
+  c = new Array(k).fill(0);
+  pathWithStartingCity = addStartingCityToPath(startingCity, intermediateCities);
+
+  // Draw graph here
+  pathData = testNewPath(cy, cy2, pathWithStartingCity, bestRouteDistance, startingCity.toString());
+  await sleep(100);
```
Comment (eAlasdair) [inline csfieldguide/static/interactives/city-trip/js/city-trip.js]: This is an interesting problem I learned about in an ENCE class. This is a delay of 0.1s *plus* the delay to run the code itself. On my low-end (but probably reasonably representative) PC it's only 40s before the interactive is an entire second behind schedule. There are ways around this but I don't think it matters enough in this situation to warrant changing anything

File: csfieldguide/static/interactives/city-trip/js/city-trip.js [hunk 15]
```diff
@@ -0,0 +1,642 @@
...
+  showTimeUnit('month', months);
+  showTimeUnit('day', days);
+  showTimeUnit('hour', hours);
+  showTimeUnit('minute', minutes);
+  showTimeUnit('second', seconds);
+}
+
+
+/** Updates the timer every second and stops the timer if time is up or path finding is stopped. */
+function startTimer(seconds) {
```
Comment (eAlasdair) [inline csfieldguide/static/interactives/city-trip/js/city-trip.js]: The timer should probably stop after reaching zero, or at least show zero at the end

File: csfieldguide/static/interactives/city-trip/js/city-trip.js [hunk 16]
```diff
@@ -0,0 +1,642 @@
...
+var stoppedMidExecution = false; // true if the stop button been clicked mid path finding.
+
+const STATUS_COMPLETE = gettext("complete!");
+const STATUS_READY = gettext("ready to go!");
+const STATUS_STOPPED = gettext("stopped");
+const STATUS_RUNNING = gettext("running");
+
+$(document).ready(function() {
+
+  updateStatus(STATUS_READY, 'status-ready');
```
Comment (eAlasdair) [inline csfieldguide/static/interactives/city-trip/js/city-trip.js]: Both maps will update, but they will be different if if the 'Add city' button is pressed after the algorithm completes
image

Comment (courtneycb) [reply csfieldguide/static/interactives/city-trip/js/city-trip.js]: Nice find!

**Comment by eAlasdair** [comment]
#381 
#722 
Appear to be relevant issues

File: csfieldguide/static/interactives/city-trip/js/city-trip.js (hunk 13 above)
Comment (courtneycb) [reply csfieldguide/static/interactives/city-trip/js/city-trip.js]: We have to fix the size of the container for the graph, so I don't think I can get it any better :( the user is able to scroll across but I know that is less than ideal. Had a talk to Jack about it.

Comment (eAlasdair) [reply csfieldguide/static/interactives/city-trip/js/city-trip.js]: Can you not have a second smaller fixed-container size for these situations? We should at least add some sort of warning like 2d-shape-manipulations has

Comment (courtneycb) [reply csfieldguide/static/interactives/city-trip/js/city-trip.js]: Added warning

File: csfieldguide/static/interactives/city-trip/js/city-trip.js [hunk 17]
```diff
@@ -0,0 +1,605 @@
+const cytoscape = require('cytoscape');
```
Comment (eAlasdair) [inline csfieldguide/static/interactives/city-trip/js/city-trip.js]: You should be able to resume path finding from where it is after the algorithm is stopped, or at least reset to the same graph

File: csfieldguide/package.json [hunk 18]
```diff
@@ -13,11 +13,13 @@
     "sticky-state": "2.4.1",
     "iframe-resizer": "3.6.3",
     "csfg-module-fsa": "file:./static/js/modules/fsa/",
+    "csfg-interactive-algorithm-timer": "file:./static/interactives/algorithm-timer/",
     "csfg-interactive-big-number-calculator": "file:./static/interactives/big-number-calculator/",
     "csfg-interactive-bin-packing": "file:./static/interactives/bin-packing/",
```
Comment (eAlasdair) [inline csfieldguide/package.json]: Dragging a box in bin-packing will also highlight text as though you were doing a click and drag to select a group of stuff

File: csfieldguide/chapters/content/en/complexity-and-tractability/sections/tractability.md [hunk 19]
```diff
@@ -37,7 +37,7 @@ It may be more meaningful to students it this is converted to some geological ti
 
 {panel end}
 
-{button-link link="http://www.csfieldguide.org.nz/releases/1.9.9/_static/widgets/tract-scaling-v2.html" text="View big time calculator"}
+{interactive slug="algorithm-timer" type="whole-page"}
```
Comment (eAlasdair) [inline csfieldguide/chapters/content/en/complexity-and-tractability/sections/tractability.md]: Perhaps remove the 2 decimal places if they aren't needed, like for a number such as 10000000000000.00

Comment (eAlasdair) [reply csfieldguide/chapters/content/en/complexity-and-tractability/sections/tractability.md]: Also how about adding order nlog(n), since that is also very common?

File: csfieldguide/package.json (hunk 18 above)
Comment (eAlasdair) [reply csfieldguide/package.json]: Perhaps also address #786 if you think it's needed

Comment (courtneycb) [reply csfieldguide/package.json]: Text highlighting error is fixed. Not sure what to add for #786 happy to hear suggestions

File: csfieldguide/static/interactives/city-trip/js/city-trip.js (hunk 17 above)
Comment (courtneycb) [reply csfieldguide/static/interactives/city-trip/js/city-trip.js]: Might add this as an issue as it will take a bit of time to implement and would rather get the chapter out

File: csfieldguide/static/interactives/algorithm-timer/js/algorithm-timer.js [hunk 20]
```diff
@@ -0,0 +1,177 @@
...
+  speed = Mathjs.bignumber(speed);
+  processors = Mathjs.bignumber(processors);
+  var steps;
+  
+  if (complexity == 'log') {
+    steps = Mathjs.bignumber(Mathjs.log(n));
+  } else if (complexity == 'linear') {
+    steps = Mathjs.bignumber(n);
+  } else if (complexity == 'nlog') {
+    steps = Mathjs.bignumber(n * Mathjs.log(n));
```
Comment (eAlasdair) [inline csfieldguide/static/interactives/algorithm-timer/js/algorithm-timer.js]: Yeah regarding earlier, these log(n) numbers should be `1 + number-rounded-up`
1 + log(1) == 1 operation
1 + log(2) == 2 operations
1 + log(3) == 2 operations
1 + log(4) == 2 operations
etc

1 + 1(log(1)) == 1 operation
1 + 2(log(2)) == 2 operations
1 + 3(log(3)) == 3 operations
etc

File: csfieldguide/static/interactives/algorithm-timer/js/algorithm-timer.js [hunk 21]
```diff
@@ -0,0 +1,177 @@
...
+  } else {
+    timeTaken = Mathjs.format(Mathjs.bignumber(timeTaken), {notation: 'fixed', precision: 2});
+  }
+
+  // If there are no significant decimal places don't show any decimal places
+  if (timeTaken % 1 == 0) {
+    timeTaken = Mathjs.format(Mathjs.bignumber(timeTaken), {notation: 'fixed', precision: 0});
+  }
+
+  if (isNaN(timeTaken)) {
```
Comment (eAlasdair) [inline csfieldguide/static/interactives/algorithm-timer/js/algorithm-timer.js]: It'd be better to find out the inputs are invalid right at the beginning, rather than after all the calculations

Comment (courtneycb) [reply csfieldguide/static/interactives/algorithm-timer/js/algorithm-timer.js]: The inputs are checked for validity before this function is called (it's not called if they are invalid). I'm not sure I can find a scenario where this if statement will be true anymore. Should I remove it or keep it there as a just-in-case thing?

File: csfieldguide/static/interactives/algorithm-timer/js/algorithm-timer.js [hunk 22]
```diff
@@ -0,0 +1,177 @@
+const Mathjs = require('mathjs');
+
```
[the remark this answers is not in the document]
Comment (eAlasdair) [reply csfieldguide/static/interactives/algorithm-timer/js/algorithm-timer.js]: Still unresolved

File: csfieldguide/static/interactives/algorithm-timer/js/algorithm-timer.js (hunk 21 above)
Comment (eAlasdair) [reply csfieldguide/static/interactives/algorithm-timer/js/algorithm-timer.js]: *Right now* it does get hit, see earlier problem. If the difference is `seconds` vs `NaN seconds` I don't think there's much benefit of keeping it just in case

File: csfieldguide/static/interactives/algorithm-timer/js/algorithm-timer.js [hunk 23]
```diff
@@ -0,0 +1,201 @@
+const Mathjs = require('mathjs');
```
[the remark this answers is not in the document]
Comment (eAlasdair) [reply csfieldguide/static/interactives/algorithm-timer/js/algorithm-timer.js]: These numbers are technically valid (from the perspective of the program) so with these values you can select different complexities and wait 5 minutes for the answer `0 seconds` to be displayed

File: csfieldguide/package.json (hunk 18 above)
Comment (eAlasdair) [inline csfieldguide/package.json]: The bin packing interactive should un-display the 'congratulations' message if the bins are packed and the user presses 'unpack all'
image

File: csfieldguide/static/interactives/algorithm-timer/js/algorithm-timer.js (hunk 23 above)
[the remark this answers is not in the document]
Comment (courtneycb) [reply csfieldguide/static/interactives/algorithm-timer/js/algorithm-timer.js]: you're very committed to your testing lol

File: csfieldguide/static/interactives/algorithm-timer/js/algorithm-timer.js [hunk 24]
```diff
@@ -0,0 +1,207 @@
...
+    $('#output').val('');
+  } else if ((1 <= n <= 1000) && !nItemsErrorMsg.hasClass('d-none')) {
+    nItems.removeClass('is-invalid');
+    nItemsErrorMsg.addClass('d-none');
+  }
+
+  // validation for speed
+  speedInput = $('#speed');
+  speedErrorMsg = $('#speed-input-error');
+  if (speed <= 0 || speed > 10000000 || isNaN(speed)) {
```
Comment (eAlasdair) [inline csfieldguide/static/interactives/algorithm-timer/js/algorithm-timer.js]: The HTML is 1 million but this is 10 million. Is that because of the multiply by ten thing? If so perhaps note it in a comment

Comment (eAlasdair) [reply csfieldguide/static/interactives/algorithm-timer/js/algorithm-timer.js]: No that is city-trip. From our discussion of this I think it should be 1 million

File: csfieldguide/package.json [hunk 25]
```diff
@@ -13,11 +13,13 @@
     "sticky-state": "2.4.1",
     "iframe-resizer": "3.6.3",
     "csfg-module-fsa": "file:./static/js/modules/fsa/",
+    "csfg-interactive-algorithm-timer": "file:./static/interactives/algorithm-timer/",
```
Comment (eAlasdair) [inline csfieldguide/package.json]: Various combinations of numbers at their highest values result in console errors
With n = 1000 there is no error but the result is still infinity

image

Comment (eAlasdair) [reply csfieldguide/package.json]: Also perhaps the number of decimal places should be limited
image

File: csfieldguide/static/interactives/city-trip/README.md [hunk 26]
```diff
@@ -0,0 +1,5 @@
+# City Trip Interactive
+
+**Author:** Courtney Bracefield, Alasdair Smith (contributed to timing logic and heapsAlgorithm.js).
+
+This interactive is created to demonstrate the travelling salesman problem.
```
Comment (eAlasdair) [inline csfieldguide/static/interactives/city-trip/README.md]: Since its for TSP perhaps it could be included again in the TSP section?

It could go just above the craypots exercise
image

**Comment by eAlasdair** [comment]
Also check #114

**Comment by courtneycb** [comment]
> Also check #114

There's not enough information in that issue to make any changes.

**Comment by eAlasdair** [comment]
> > Also check #114
> 
> There's not enough information in that issue to make any changes.

Then just close it?

**Comment by courtneycb** [comment]
> > > Also check #114
> > 
> > 
> > There's not enough information in that issue to make any changes.
> 
> Then just close it?

Should probably get someone who was around back then (i.e Jack) to double check before closing

**Comment by eAlasdair** [comment]
> Should probably get someone who was around back then (i.e Jack) to double check before closing

Oh, I did close a couple of similar ones... But its so unspecific and out of date, it can be remade if need be

File: csfieldguide/static/interactives/big-number-calculator/js/big-number-calculator.js [hunk 27]
```diff
@@ -3,106 +3,110 @@
...
 
 $(document).ready(function () {
-    $('#interactive-big-number-calculator button').on('click', function(){
-        var button_type = $(this).attr('id').split('-').pop();
-        var x = getXValue();
-        var y = getYValue();
+  $('#interactive-big-number-calculator button').on('click', function(){
+    var button_type = $(this).attr('id').split('-').pop();
+    var x = getXValue();
+    var y = getYValue();
```
Comment (eAlasdair) [inline csfieldguide/static/interactives/big-number-calculator/js/big-number-calculator.js]: Why the change from 4-space indents to 2-space? I thought 4 was the standard for JS?
Now that I look at it *almost* everything is 2-space, should I use that in my work?

File: csfieldguide/templates/interactives/algorithm-timer.html [hunk 28]
```diff
@@ -0,0 +1,191 @@
+{% extends interactive_mode_template %}
```
Comment (eAlasdair) [inline csfieldguide/templates/interactives/algorithm-timer.html]: For really large results the 2 decimal places are still shown (regardless of unit)
image

File: csfieldguide/tests/chapters/loaders/test_chapters_loader.py [hunk 29]
```diff
@@ -118,7 +118,8 @@ def test_chapters_chapter_loader_interactive(self, find_image_files):
             structure_filename="{}.yaml".format(test_slug)
         )
         chapter_loader.load()
-        self.assertTrue(find_image_files.called)
+        # To be added back in. See issue/919
+        # self.assertTrue(find_image_files.called)
```
Comment (eAlasdair) [inline csfieldguide/tests/chapters/loaders/test_chapters_loader.py]: This stuff still needs to be reverted

File: csfieldguide/static/interactives/big-number-calculator/js/big-number-calculator.js (hunk 27 above)
Comment (courtneycb) [reply csfieldguide/static/interactives/big-number-calculator/js/big-number-calculator.js]: We follow Google's Javascript Style Guide which has 2-space indentation for JS. See https://google.github.io/styleguide/jsguide.html#formatting-block-indentation

File: csfieldguide/templates/interactives/algorithm-timer.html (hunk 28 above)
Comment (eAlasdair) [reply csfieldguide/templates/interactives/algorithm-timer.html]: As an aside, this is the largest possible result
image

**Comment by courtneycb** [comment]
@eAlasdair ready for re-review

File: csfg [hunk 30]
```diff
@@ -83,7 +83,6 @@ cmd_update() {
   echo ""
   cmd_migrate
 
-  echo ""
```
Comment (eAlasdair) [inline csfg]: I assume this is fine? Strange to see an edit to csfg in this PR

File: csfieldguide/utils/check_required_files.py [hunk 31]
```diff
@@ -27,7 +27,8 @@ def find_image_files(images, md_file_path):
         CouldNotFindImageError: when image file cannot be found.
     """
     for image in images:
-        if not finders.find(image):
+        # Need to check interactives folder eventually. See issue/919
+        if not image.startswith('interactives/') and not finders.find(image):
             raise CouldNotFindImageError(image, md_file_path)
```
Comment (eAlasdair) [inline csfieldguide/utils/check_required_files.py]: Needs to be reverted, unless there's a reason for this
